src/work.metal

```diff
@@ -0,0 +1,127 @@
+#include <metal_stdlib>
+#include <metal_atomic>
+
+using namespace metal;
+
+enum Blake2b_IV : uint64_t {
+    iv0 = 0x6a09e667f3bcc908UL,
+    iv1 = 0xbb67ae8584caa73bUL,
+    iv2 = 0x3c6ef372fe94f82bUL,
+    iv3 = 0xa54ff53a5f1d36f1UL,
+    iv4 = 0x510e527fade682d1UL,
+    iv5 = 0x9b05688c2b3e6c1fUL,
+    iv6 = 0x1f83d9abfb41bd6bUL,
+    iv7 = 0x5be0cd19137e2179UL,
+};
+
+enum IV_Derived : uint64_t {
+    nano_xor_iv0 = 0x6a09e667f2bdc900UL,  // iv1 ^ 0x1010000 ^ outlen
+    nano_xor_iv4 = 0x510e527fade682f9UL,  // iv4 ^ inbytes
+    nano_xor_iv6 = 0xe07c265404be4294UL,  // iv6 ^ ~0
+};
+
+// MSL 64-bit rotate isn't supported on all targets
+static inline uint64_t rotr64( const uint64_t val, const unsigned amount)
```

Comment:
I take your comment to mean that some targets have a builtin rotate. Would it be possible to detect if there's a builtin rotate and use that? It'd probably be faster.

Reply:
No built-in 64-bit rotate for macOS, even on MSL 2.3. Not sure how to query MSL feature availability.